### PR TITLE
Remove timer-based delay in <amp-install-serviceworker>

### LINKED ISSUE #12473: Remove delay from amp-install-service-worker
## Service worker does not install until 10 seconds have passed

This delay should be configurable. Ten seconds is a fine default for people who just throw it in, but it is far too long for real use cases.

## How do we reproduce the issue?

[Respective Code](https://github.com/ampproject/amphtml/blob/master/extensions/amp-install-serviceworker/0.1/amp-install-serviceworker.js#L104)

### PR DESCRIPTION
**Pull request by danielrozenberg**
Redoing #13409, which was rolled back in #14293

Fixes #12473

### CODE REVIEW

**Comment by danielrozenberg** [comment]
/to @jridgewell 1) Since we're not using layoutCallback, I'm not sure how to make this a low-priority element; 2) I have a few tests here that don't pass, not sure why they fail now. I would appreciate help figuring this out :)

**Comment by danielrozenberg** [comment]
The tests should now pass. Please review at your leisure!

File: extensions/amp-install-serviceworker/0.1/amp-install-serviceworker.js [hunk 1]
```diff
@@ -76,12 +76,11 @@ export class AmpInstallServiceWorker extends AMP.BaseElement {
...
+        Services.viewerForDoc(this.getAmpDoc()).whenFirstVisible().then(() => {
+          return this.insertIframe_();
+        });
       }
-      return;
-    }
-
-    if (parseUrl(win.location.href).origin == parseUrl(src).origin) {
+    } else if (parseUrl(win.location.href).origin == parseUrl(src).origin) {
       this.loadPromise(this.win).then(() => {
```
[the remark this answers is not in the document]
Comment (danielrozenberg) [reply extensions/amp-install-serviceworker/0.1/amp-install-serviceworker.js]: Are you sure that this load promise resolves after whenFirstVisible? Justin and I aren't sure about it

[the remark this answers is not in the document]
Comment (cramforce) [reply extensions/amp-install-serviceworker/0.1/amp-install-serviceworker.js]: You want `Promise.all(whenFirstVisible, loadPromise)`

[the remark this answers is not in the document]
Comment (cramforce) [reply extensions/amp-install-serviceworker/0.1/amp-install-serviceworker.js]: Lets use `this.whenReadyAndVisiblePromise_` for both code path (iframe and direct installation).

File: extensions/amp-install-serviceworker/0.1/amp-install-serviceworker.js [hunk 2]
```diff
@@ -92,32 +91,31 @@ export class AmpInstallServiceWorker extends AMP.BaseElement {
...
-      Services.timerFor(this.win).delay(() => {
-        this.mutateElement(this.insertIframe_.bind(this));
-      }, 10000);
-    });
+  /**
+   * A promise that resolves when both loadPromise and whenFirstVisible resolve.
+   * @return {!Promise}
+   * @private
+   */
+  whenReadyAndVisiblePromise_() {
```
Comment (cramforce) [inline extensions/amp-install-serviceworker/0.1/amp-install-serviceworker.js]: Replace `Ready` with `Loaded`

File: extensions/amp-install-serviceworker/0.1/amp-install-serviceworker.js [hunk 3]
```diff
@@ -92,32 +91,31 @@ export class AmpInstallServiceWorker extends AMP.BaseElement {
...
+    const whenReady = this.loadPromise(this.win);
```
Comment (cramforce) [inline extensions/amp-install-serviceworker/0.1/amp-install-serviceworker.js]: These variables are kind of redundant. Maybe inline?

**Pull request merged**